File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution {
    public int minSetSize(int[] arr) {
        int n = arr.length;
        Map<Integer, Integer> map = new HashMap<>();
        for(int a: arr){
            map.put(a, map.getOrDefault(a, 0) + 1);
        }
        int values[] = new int[map.size()];
        int idx = 0;
        for(int a: map.values()){
            values[idx] = a;
            idx++;
        }
        Arrays.sort(values);
        int count = 0;
        for(int i = values.length-1; i>=0; i--){
            count += values[i];
            if(count >= n/2)
                return values.length-i;
        }
        
        return -1;
    }
}

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.py
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        m={}
        for i in arr:
            if i not in m:
                m[i]=1
            else:
                m[i]+=1

        s=[]
        for i in m:
            s.append(m[i])
        s.sort(reverse=True)
        count=0
        ans=0
        i=0
        while count<len(arr)//2:
            count+=s[i]
            i+=1
            ans+=1
        return ans